src/platform/OpenThread/GenericNetworkCommissioningThreadDriver.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/NetworkCommissioning.h>
+#include <platform/OpenThread/GenericNetworkCommissioningThreadDriver.h>
+#include <platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h>
+#include <platform/ThreadStackManager.h>
+
+#include <limits>
+
+using namespace chip;
+using namespace chip::Thread;
+
+namespace chip {
+namespace DeviceLayer {
+namespace NetworkCommissioning {
+// NOTE: For ThreadDriver, we uses two network configs, one is mSavedNetwork, and another is mStagingNetwork, during init, it will
+// load the network config from thread persistent info, and loads it into both mSavedNetwork and mStagingNetwork. When updating the
+// networks, all changed are made on the staging network. When validated we can commit it and save it to the persisten info
+
+CHIP_ERROR GenericThreadDriver::Init()
+{
+    ByteSpan currentProvision;
+    VerifyOrReturnError(ThreadStackMgrImpl().IsThreadAttached(), CHIP_NO_ERROR);
+    VerifyOrReturnError(ThreadStackMgrImpl().GetThreadProvision(currentProvision) == CHIP_NO_ERROR, CHIP_NO_ERROR);
+
+    mSavedNetwork.Init(currentProvision);
+    mStagingNetwork.Init(currentProvision);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericThreadDriver::CommitConfiguration()
+{
+    // Note: on AttachToThreadNetwork OpenThread will persist the networks by their own,
+    // we don't have much to do for saving the networks (see Init() above,
+    // we just loads the saved dataset from ot instance.)
+    mSavedNetwork = mStagingNetwork;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericThreadDriver::RevertConfiguration()
+{
+    mStagingNetwork = mSavedNetwork;
+    return CHIP_NO_ERROR;
+}
+
+Status GenericThreadDriver::AddOrUpdateNetwork(ByteSpan operationalDataset)
+{
+    uint8_t extpanid[kSizeExtendedPanId];
+    uint8_t newExtpanid[kSizeExtendedPanId];
+    Thread::OperationalDataset newDataset;
+
+    newDataset.Init(operationalDataset);
+    VerifyOrReturnError(newDataset.IsCommissioned(), Status::kOutOfRange);
+
+    VerifyOrReturnError(!mStagingNetwork.IsCommissioned() || memcmp(extpanid, newExtpanid, kSizeExtendedPanId) == 0,
+                        Status::kBoundsExceeded);
```

Comment:
Why this status?  Seems odd.

Reply:
I added a missing step and comment to explain hope this makes more sense. With thread, we only maintain one active operational dataset. So If the StagingNetwork is already in use (commissioned), we will only update it within the same extended pan id.